[macOS Ventura] Build fails when using public SDK

emw-apple opened the pull request:
#### 16213795b52db0b2171464a228870d8fb84627c9
<pre>
[macOS Ventura] Build fails when using public SDK
<a href="https://bugs.webkit.org/show_bug.cgi?id=242429">https://bugs.webkit.org/show_bug.cgi?id=242429</a>

Reviewed by Tim Horton.

These changes should be enough to get the Mac build succeeding with the
latest Xcode 14 beta and macOS Venture SDK.

Unless otherwise noted, all changes are declaring and updating SPI.

* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::retry): ApplePayAuthorizationResult
  may not have an orderDetails member. Avoid designated initialization
  to simplify the use site.
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h: Import public request
  status API for PASSKIT_PAYMENT_ORDER_DETAILS feature, along with
  declaring new SPI. Additionally, begin importing the
  &lt;PassKit/PassKit.h&gt; umbrella header on macOS, and clean up SPI
  declarations that have become API since Big Sur.
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/VisionKitCoreSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServicePickerSPI.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FullscreenVideoTextRecognition.mm:

Canonical link: <a href="https://commits.webkit.org/253305@main">https://commits.webkit.org/253305@main</a>
</pre>

aproskuryakov commented on Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
```diff
@@ -71,6 +71,12 @@
+#import <PassKitCore/PKPaymentRequestStatus.h>
+#else
+#import <PassKit/PKPaymentRequestStatus.h>
+#endif
```
This says that we import `PassKitCore/PKPaymentRequestStatus.h` from macOS 13 internal SDK, and `PassKit/PKPaymentRequestStatus.h` in all other cases. And it's based on target OS, not on SDK version.

This feels very unusual, is this indeed the intention?

emw-apple replied:
Not intentional. Looks like I should follow the above convention, and use:
```
#if HAVE(PASSKIT_PAYMENT_ORDER_DETAILS)
#if HAVE(PASSKIT_MODULARIZATION) && USE(APPLE_INTERNAL_SDK)
```

aproskuryakov commented on Source/WebCore/PAL/pal/spi/mac/VisionKitCoreSPI.h:
```diff
@@ -0,0 +1,55 @@
+
+#pragma once
+
+#import <AppKit/AppKit.h>
```
Will we need a separate SPI header for iOS? I don't know much about VisionKit, but I'd expect it to be similar between platforms.

whsieh replied:
Yeah, I don't think we should try to maintain separate `VisionKitCoreSPI.h` SPI headers on macOS and iOS.

Maybe we should including some pieces of `VKDefines.h`? e.g.

```
#if PLATFORM(IOS)
#import <UIKit/UIKit.h>
typedef UIImage VKImageClass;
typedef UIResponder VKResponder;
typedef UIView VKBaseViewClass;
typedef UIWindow VKWindowClass;
#elif PLATFORM(MAC)
#import <AppKit/AppKit.h>
typedef NSImage VKImageClass;
typedef NSResponder VKResponder;
typedef NSView VKBaseViewClass;
typedef NSWindow VKWindowClass;
#endif
```

(Platform-specific API, such as `VKCImageAnalysisOverlayView`, would then be behind the appropriate compile-time guards such as `PLATFORM(MAC)`).

whsieh replied:
Ah, I see now — these macOS/iOS-specific headers only include API that's specific to those platforms...

In any case, I think it would probably be best to have all the declarations available via a single header, such that there's a single `VisionKitCoreSPI.h` header to import from PAL that contains all the VKC symbols.

emw-apple replied:
Makes sense! I don't think we need the full expressivity of `VKDefines` at this point (maybe I'll discover that we do when I work on the open source iOS build…), since all we need at this time is the Mac-specific `VKCImageAnalysisOverlayView`.

aproskuryakov commented on Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
```diff
@@ -132,6 +132,7 @@
 #import <pal/spi/mac/NSTextInputContextSPI.h>
 #import <pal/spi/mac/NSViewSPI.h>
 #import <pal/spi/mac/NSWindowSPI.h>
+#import <pal/spi/mac/VisionKitCoreSPI.h>
```
This looks like it would break iOS build. The other `pal/spi/mac` headers included here have `#if PLATFORM(MAC)` or `#if USE(APPKIT)`.

emw-apple replied:
This whole file is Mac-only, it's for the `WebView` NSView subclass.

aproskuryakov replied:
I see. What a brilliant location for a Mac only file.

dcrousso commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -888,7 +888,11 @@ ExceptionOr<void> ApplePayPaymentHandler::retry(PaymentValidationErrors&& valida
 
+#if ENABLE(APPLE_PAY_PAYMENT_ORDER_DETAILS)
+    ApplePayPaymentAuthorizationResult authorizationResult { ApplePayPaymentAuthorizationResult::Failure, WTFMove(errors), std::nullopt };
+#else
```
Rather than repeat this twice, how about we separate this onto separate lines as that would let us avoid having to even care about `ENABLE(APPLE_PAY_PAYMENT_ORDER_DETAILS)` entirely?
```
ApplePayPaymentAuthorizationResult authorizationResult;
authorizationResult.status = ApplePayPaymentAuthorizationResult::Failure;
authorizationResult.errors = WTFMove(errors);
```

dcrousso commented on Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
```diff
@@ -71,6 +71,14 @@
+#if HAVE(PASSKIT_MODULARIZATION) && USE(APPLE_INTERNAL_SDK)
+#import <PassKitCore/PKPaymentRequestStatus.h>
+#else
+#import <PassKit/PKPaymentRequestStatus.h>
```
this should already be handled by `#import <PassKit/PassKit.h>`.  is that not working?

also, the new `PKPaymentOrderDetails` is public API, so not really sure why we'd need to do anything special for it (i.e. this file is for handling SPI that public SDKs dont have, not API that they do)

emw-apple replied:
The immediate challenge is that the `<PassKit/PassKit.h>` import happens inside of an `#if PLATFORM(IOS_FAMILY)` block. I tried moving it out, and it revealed other duplicate declarations that would need cleanup, so went with this approach to keep the scope of this PR small.

Happy to try again, as it does feel like the right thing to do.

emw-apple replied:
Tried again, and was able to clean up a lot more, by relying on the umbrella `PassKit.h` import more and removing declarations we only needed for pre-Big Sur OSes. Let me know how that looks!

dcrousso commented on Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
```diff
@@ -531,6 +557,7 @@ typedef NS_ENUM(NSInteger, PKPaymentButtonType) {
 
 NS_ASSUME_NONNULL_BEGIN
 
+#if !HAVE(PASSKIT_PAYMENT_ORDER_DETAILS)
```
i dont think this is the right change, as `PKPaymentAuthorizationResult` is not dependent on `HAVE(PASSKIT_PAYMENT_ORDER_DETAILS)`.  it predates it

what exactly is the error that is necessitating this change?

emw-apple replied:
`#import <PassKit/PKPaymentRequestStatus.h>` on L78 is gated by `HAVE(PASSKIT_PAYMENT_ORDER_DETAILS)`, and it pulls PassKit's declaration of these classes. So I wanted to continue to declare our SPI when PASSKIT_PAYMENT_ORDER_DETAILS is off (effectively, in Monterey / iOS 15 and below).

Looking back at this, though, it seems like these types may have been public API in previous releases, and we can clean them up further.

emw-apple commented on Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
```diff
@@ -409,6 +409,7 @@ extern CGDataProviderRef CGDataProviderCreateMultiRangeDirectAccess(
 #if HAVE(LOCKDOWN_MODE_PDF_ADDITIONS)
 CG_EXTERN void CGEnterLockdownModeForPDF();
 CG_LOCAL bool CGIsInLockdownModeForPDF();
+CG_EXTERN void CGEnterLockdownModeForFonts();
```
CC @brentfulgham. As far as I can tell, this needs to be added in addition to the Lockdown Mode SPI we landed in #2426. Perhaps it shouldn't be under `LOCKDOWN_MODE_PDF_ADDITIONS`, but we don't appear to have a fonts-specific PlatformHave macro.

hortont424 replied:
I don't really think SPI headers *need* #ifs.

hortont424 reviewed (approved):
Shocked and delighted by how red this patch is.

aproskuryakov commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -888,7 +888,9 @@ ExceptionOr<void> ApplePayPaymentHandler::retry(PaymentValidationErrors&& valida
         errors.append(ApplePayError::create(ApplePayErrorCode::Unknown, std::nullopt, nullString()));
 
-    ApplePayPaymentAuthorizationResult authorizationResult { ApplePayPaymentAuthorizationResult::Failure, WTFMove(errors) };
+    ApplePayPaymentAuthorizationResult authorizationResult;
```
The downside is that if more fields are added, we won't know that they are uninitialized here. I think that I'd duplicate the whole long line with `ENABLE(APPLE_PAY_PAYMENT_ORDER_DETAILS)` ifdef to protect against that.

emw-apple committed:
[macOS Ventura] Build fails when using public SDK
https://bugs.webkit.org/show_bug.cgi?id=242429

Reviewed by Tim Horton.

These changes should be enough to get the Mac build succeeding with the
latest Xcode 14 beta and macOS Venture SDK.

Unless otherwise noted, all changes are declaring and updating SPI.

* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::retry): ApplePayAuthorizationResult
  may not have an orderDetails member. Avoid designated initialization
  to simplify the use site.
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h: Import public request
  status API for PASSKIT_PAYMENT_ORDER_DETAILS feature, along with
  declaring new SPI. Additionally, begin importing the
  <PassKit/PassKit.h> umbrella header on macOS, and clean up SPI
  declarations that have become API since Big Sur.
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/VisionKitCoreSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServicePickerSPI.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FullscreenVideoTextRecognition.mm:

Canonical link: https://commits.webkit.org/253305@main
